### PR TITLE
Use own fifo library

### PR DESCRIPTION
**Pull request by m-mizutani**
## Background

- `ruby-fifo` gem does not handle return value from File.read properly. Then it raises Exception and stop if File.read receives EOF
- `ruby-fifo` read from named pipe byte by byte. It may cause performance issue
- The gem looks like not maintained

## Changes

- Create `lib/fluent/plugin/fifo.rb` as own fifo library
- Use`IO.sysread` instead of `File.read` to avoid read blocking
- Add error handling for EOF

### CODE REVIEW

**Comment by sonots** [comment]
Nice. I wanted to graduate from ruby-fifo gem https://github.com/sonots/fluent-plugin-named_pipe/issues/6. I will merge after tests passed.

File: fluent-plugin-named_pipe.gemspec [hunk 1]
```diff
@@ -17,7 +17,6 @@ Gem::Specification.new do |s|
   s.require_paths = ["lib"]
 
   s.add_runtime_dependency "fluentd", ">= 0.10.58"
-  s.add_runtime_dependency "ruby-fifo", "0.0.1"
   if RUBY_PLATFORM =~ /mswin|mingw/i
     s.add_runtime_dependency "win32-pipe"
```
Comment (sonots) [inline fluent-plugin-named_pipe.gemspec]: Maybe, we can drop windows support

Comment (m-mizutani) [reply fluent-plugin-named_pipe.gemspec]: Agree, and my changes support only POSIX environment now...

**Comment by sonots** [comment]
Hmm, travis is failing like:

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

**Comment by m-mizutani** [comment]
I think it’s caused by read blocking when plugin shutdown. Let me change shutdown procedure to avoid blocking.

**Comment by m-mizutani** [comment]
I updated my PR.

- Implement graceful shutdown without read blocking by IO.select and timeout
- Remove Windows support part

**Comment by m-mizutani** [comment]
@sonots 
Sorry for bothering you, but I appreciate you can review and merge this one. Thank you.

**Comment by sonots** [comment]
Thanks, I will look later.

**Comment by sonots** [comment]
LGTM!

**Comment by sonots** [comment]
0.2.0 is released.

**Comment by m-mizutani** [comment]
Thank you!

**Pull request merged**